eos-updater-avahi/docs/eos-updater-avahi.8
.\" Manpage for eos-updater-avahi.
.\" SPDX-License-Identifier: CC-BY-SA-4.0
.\" SPDX-FileCopyrightText: Copyright © 2017 Endless Mobile, Inc.
.TH man 8 "20 Feb 2017" "1.0" "eos\-updater\-avahi man page"
.\"
.SH NAME
.IX Header "NAME"
eos\-updater\-avahi — Endless OS Avahi Advertisement Updater
.\"
.SH SYNOPSIS
.IX Header "SYNOPSIS"
.\"
\fBeos\-updater\-avahi [\-q] [\-d \fPservice\-directory\fB]
[\-c \fPconfig\-file\fB]
.\"
.SH DESCRIPTION
.IX Header "DESCRIPTION"
.\"
\fBeos\-updater\-avahi\fP is a utility to update the Avahi advertisements for
Endless OS updates being advertised by this computer to others on the local
network. When run, it checks whether advertising updates is enabled, and which
OSTree commits are present in the computer’s repository, and adds, updates or
removes the Avahi advertisements for them.
.PP
The Avahi advertisements are DNS\-SD records which allow other computers on the
local network to check which updates are available to download from this
computer, and which port to download them from. By default, they are stored in
\fI/etc/avahi/services/eos\-ostree\-updater\-*.service\fP.
.PP
Advertisement of updates can be disabled by setting the \fIAdvertiseUpdates=\fP
key to \fIfalse\fP in \fBeos\-update\-server.conf\fP(5).
.PP
\fBeos\-updater\-avahi\fP is designed to be run by \fBsystemd\fP(1), whenever
the \fIeos\-updater\-avahi.path\fP unit is triggered. By default, it is
triggered whenever the system OSTree repository is updated, or the updater
configuration is changed. There should be no need to run
\fBeos\-updater\-avahi\fP manually.
.PP
If the computer has been converted to not use OSTree, advertisement of updates
to the local network is permanently disabled.
.\"
.SH ROUTER CONFIGURATION
.IX HEADER "ROUTER CONFIGURATION"
.\"
\fBeos\-updater\-avahi\fP relies on DNS\-SD advertisements being successfully
multicast across the local network, via mDNS over IPv4 or IPv6 (or both). As
multicast is expensive on wireless networks, some routers do not forward
multicast packets to wireless peers unless they can determine that the peers are
interested in receiving them. This is typically achieved by listening for IGMP
requests from the peers to establish their interest in receiving multicast
packets.
.PP
If DNS\-SD advertisement of updates in the local network is not working, or if
it works when a peer is first started, but then times out after a number of
minutes, please ensure that the relevant multicast options are enabled in the
local network’s router. These might be labelled \fIIGMP snooping\fP,
\fIIGMP proxy\fP or \fImulticast routing\fP, for example.
.PP
mDNS problems can be debugged using \fBavahi\-browse\fP(1).
.\"
.SH OPTIONS
.IX Header "OPTIONS"
.\"
.IP "\fB\-q\fP, \fB\-\-quiet\fP"
Do not print informational output.
.\"
.IP "\fB\-d\fP, \fB\-\-service\-directory=\fP"
Look for and modify Avahi service files in the given directory. This is
intended to be used for testing. (Default: \fI/etc/avahi/services\fP.)
.\"
.IP "\fB\-c\fP, \fB\-\-config\-file=\fP"
Load configuration for \fBeos\-updater\-avahi\fP from the given file. This is
intended to be used for testing. (Default:
\fI/etc/eos\-updater/eos\-updater\-avahi.conf\fP.)
.\"
.SH "ENVIRONMENT"
.IX Header "ENVIRONMENT"
.\"
\fPeos\-updater\-avahi\fP supports the standard GLib environment variables for
debugging. These variables are \fBnot\fP intended to be used in production:
.\"
.IP \fI$G_MESSAGES_DEBUG\fP 4
.IX Item "$G_MESSAGES_DEBUG"
This variable can contain one or more debug domain names to display debug output
for. The value \fIall\fP will enable all debug output. The default is for no
debug output to be enabled.
.\"
.SH "EXIT STATUS"
.IX Header "EXIT STATUS"
.\"
\fBeos\-updater\-avahi\fP may return one of several error codes if it encounters
problems.
.\"
.IP "0" 4
.IX Item "0"
No problems occurred. The Avahi service file was successfully updated if
necessary.
.\"
.IP "1" 4
.IX Item "1"
There was an error while updating or deleting the Avahi service file.
.\"
.IP "2" 4
.IX Item "2"
An invalid option was passed to \fBeos\-updater\-avahi\fP on startup.
.\"
.IP "3" 4
.IX Item "3"
The configuration file for \fBeos\-updater\-avahi\fP could not be loaded or
parsed.
.\"
.SH "FILES"
.IX Header "FILES"
.\"
.IP \fI/etc/avahi/services/eos\-ostree\-updater\-[0–65535].service\fP 4
.IX Item "/etc/avahi/services/eos-ostree-updater\-[0–65535].service"
Describes the DNS\-SD record for Avahi to advertise to the local network.
Many files can exist, one for each advertised repository. No advertisement will
be made if none of these files are present. These files are managed by
\fBeos\-updater\-avahi\fP.
.\"
.IP \fI/etc/eos\-updater/eos\-update\-server.conf\fP 4
.IX Item "/etc/eos\-updater/eos\-update\-server.conf"
.IX Item "/usr/local/share/eos\-updater/eos\-update\-server.conf"
.IX Item "/usr/share/eos\-updater/eos\-update\-server.conf"
Configuration file for \fBeos\-updater\-avahi\fP. Defaults are loaded from
\fI/usr/local/share/eos\-updater/eos\-update\-server.conf\fP (if present) or
\fI/usr/share/eos\-updater/eos\-update\-server.conf\fP. See
\fBeos\-update\-server.conf\fP(5).
.\"
.IP \fI/lib/systemd/system/eos\-updater\-avahi.path\fP 4
.IX Item "/lib/systemd/system/eos\-updater\-avahi.path"
\fBsystemd\fP(1) path file which triggers \fBeos\-updater\-avahi\fP to run
when conditions change. See \fBsystemd.path\fP(5).
.\"
.IP \fI/lib/systemd/system/eos\-updater\-avahi.service\fP 4
.IX Item "/lib/systemd/system/eos\-updater\-avahi.service"
\fBsystemd\fP(1) service file which specifies the runtime environment for
\fBeos\-updater\-avahi\fP. See \fBsystemd.service\fP(5).
.\"
.SH "SEE ALSO"
.IX Header "SEE ALSO"
.\"
\fBeos\-update\-server\fP(8),
\fBeos\-updater\fP(8),
\fBeos\-update\-server.conf\fP(5)
.PP
\fBavahi\-browse\fP(1)
.PP
\fIhttps://en.wikipedia.org/wiki/DNS\-SD\fP,
\fIhttps://en.wikipedia.org/wiki/Multicast_DNS\fP,
\fIhttps://en.wikipedia.org/wiki/Internet_Group_Management_Protocol\fP
.\"
.SH BUGS
.IX Header "BUGS"
.\"
Any bugs which are found should be reported on the project website:
.br
\fIhttps://support.endlessm.com/\fP
.\"
.SH AUTHOR
.IX Header "AUTHOR"
.\"
Endless Mobile, Inc.
.\"
.SH COPYRIGHT
.IX Header "COPYRIGHT"
.\"
Copyright © 2017 Endless Mobile, Inc.


debian/changelog
eos-updater (1.3.0~rc3) unstable; urgency=low

  * Updating version to 1.3.0~rc3

 -- Roddy Shuler <roddy@endlessm.com>  Tue, 12 Nov 2013 21:13:35 -0800

eos-updater (1.3.0~rc1) unstable; urgency=low

  * Updating version to 1.3.0~rc1

 -- Roddy Shuler <roddy@endlessm.com>  Fri, 08 Nov 2013 23:38:59 -0800

eos-updater (1.2.1~rc1) unstable; urgency=low

  * Updating version to 1.2.1~rc1

 -- Roddy Shuler <roddy@endlessm.com>  Sun, 27 Oct 2013 22:30:28 -0700

eos-updater (0-1) UNRELEASED; urgency=low

  * Initial release (Closes: #1144)

 -- John Hiesey <john@endlessm.com>  Fri, 04 Oct 2013 11:44:06 -0700


eos-updater-avahi/eos-updater-avahi.c
/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
 *
 * Copyright © 2017 Endless Mobile, Inc.
 *
 * SPDX-License-Identifier: LGPL-2.1-or-later
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * Authors:
 *  - Philip Withnall <withnall@endlessm.com>
 *  - Krzesimir Nowak <krzesimir@kinvolk.io>
 */

#include <glib.h>
#include <libeos-update-server/config.h>
#include <libeos-updater-util/avahi-service-file.h>
#include <libeos-updater-util/ostree-util.h>
#include <libeos-updater-util/util.h>
#include <locale.h>
#include <ostree.h>
#include <stdlib.h>

static gboolean
get_refs (OstreeRepo            *repo,
          OstreeCollectionRef ***out_refs,
          GCancellable          *cancellable,
          GError               **error)
{
  g_autoptr(GHashTable) refs = NULL;
  GHashTableIter iter;
  gpointer key, value;
  g_autoptr(GPtrArray) refs_array = NULL;

  if (!ostree_repo_list_collection_refs (repo,
                                         /* match_collection_id: */ NULL,
                                         &refs,
                                         OSTREE_REPO_LIST_REFS_EXT_NONE,
                                         cancellable,
                                         error))
    return FALSE;

  if (g_hash_table_size (refs) == 0)
    {
      g_set_error_literal (error, G_IO_ERROR, G_IO_ERROR_FAILED,
                           "no refs to advertise");
      return FALSE;
    }
  refs_array = g_ptr_array_new_full (g_hash_table_size (refs) + 1, (GDestroyNotify)ostree_collection_ref_free);
  g_hash_table_iter_init (&iter, refs);
  while (g_hash_table_iter_next (&iter, &key, &value))
    {
      g_autoptr(OstreeCollectionRef) ref = key;
      g_autofree gchar *checksum = value;

      g_hash_table_iter_steal (&iter);
      g_ptr_array_add (refs_array, g_steal_pointer (&ref));
    }
  g_ptr_array_add (refs_array, NULL);

  g_assert (out_refs);
  *out_refs = (OstreeCollectionRef **)g_ptr_array_free (g_steal_pointer (&refs_array), FALSE);
  return TRUE;
}

static gboolean
get_raw_summary_timestamp_from_metadata (GBytes    *summary,
                                         gboolean  *out_found,
                                         guint64   *out_timestamp,
                                         GError   **error)
{
  g_autoptr(GVariant) summary_variant = g_variant_ref_sink (g_variant_new_from_bytes (OSTREE_SUMMARY_GVARIANT_FORMAT, summary, FALSE));
  g_autoptr(GVariant) additional_metadata = NULL;
  g_autoptr(GVariant) value = NULL;

  if (!g_variant_is_normal_form (summary_variant))
    {
      *out_found = FALSE;
      *out_timestamp = 0;
      g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND,
                   "Corrupt summary file");
      return FALSE;
    }

  additional_metadata = g_variant_get_child_value (summary_variant, 1);
  value = g_variant_lookup_value (additional_metadata, "ostree.summary.last-modified", G_VARIANT_TYPE_UINT64);

  g_assert (out_found != NULL);
  g_assert (out_timestamp != NULL);
  if (value == NULL)
    {
      *out_found = FALSE;
      *out_timestamp = 0;
      return TRUE;
    }

  /* FIXME: Disable diagnostics for GUINT64_TO_BE():
   * https://bugzilla.gnome.org/show_bug.cgi?id=788384 */
_Pragma ("GCC diagnostic push")
_Pragma ("GCC diagnostic ignored \"-Wsign-conversion\"")
  *out_found = TRUE;
  *out_timestamp = GUINT64_FROM_BE (g_variant_get_uint64 (value));
_Pragma ("GCC diagnostic pop")
  return TRUE;
}

static gboolean
bad_timestamp (guint64   modification_time_secs,
               GError  **error)
{
  g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
               "Invalid timestamp %" G_GUINT64_FORMAT,
               modification_time_secs);
  return FALSE;
}

static gboolean
get_summary_timestamp_from_guint64 (guint64     modification_time_secs,
                                    GDateTime **out_summary_timestamp,
                                    GError    **error)
{
  g_autoptr(GDateTime) summary_timestamp = NULL;

  if (modification_time_secs > G_MAXINT64)
    return bad_timestamp (modification_time_secs, error);

  summary_timestamp = g_date_time_new_from_unix_utc ((gint64)modification_time_secs);
  if (summary_timestamp == NULL)
    return bad_timestamp (modification_time_secs, error);

  g_assert (out_summary_timestamp);
  *out_summary_timestamp = g_steal_pointer (&summary_timestamp);
  return TRUE;
}

static gboolean
get_summary_timestamp (OstreeRepo    *repo,
                       GDateTime    **out_summary_timestamp,
                       GCancellable  *cancellable,
                       GError       **error)
{
  g_autoptr(GFile) summary_file = g_file_get_child (ostree_repo_get_path (repo), "summary");
  guint64 raw_timestamp;
  gboolean found = FALSE;
  g_autoptr(GBytes) summary_bytes = NULL;

  if (!eos_updater_read_file_to_bytes (summary_file, cancellable, &summary_bytes, error))
    return FALSE;

  if (summary_bytes != NULL)
    {
      if (!get_raw_summary_timestamp_from_metadata (summary_bytes, &found, &raw_timestamp, error))
        return FALSE;
    }

  if (!found)
    {
      g_autoptr(GFileInfo) summary_info = NULL;
      g_autoptr(GError) local_error = NULL;

      summary_info = g_file_query_info (summary_file,
                                        G_FILE_ATTRIBUTE_TIME_MODIFIED,
                                        G_FILE_QUERY_INFO_NONE,
                                        cancellable,
                                        &local_error);
      if (summary_info == NULL)
        {
          if (!g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
            {
              g_propagate_error (error, g_steal_pointer (&local_error));
              return FALSE;
            }
          g_clear_error (&local_error);
        }
      else
        {
          raw_timestamp = g_file_info_get_attribute_uint64 (summary_info,
                                                            G_FILE_ATTRIBUTE_TIME_MODIFIED);
          found = TRUE;
        }
    }

  if (found)
    return get_summary_timestamp_from_guint64 (raw_timestamp, out_summary_timestamp, error);

  g_set_error (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND, "No summary file found");
  return FALSE;
}

static gboolean
get_refs_and_summary_timestamp (OstreeSysroot         *sysroot,
                                OstreeCollectionRef ***out_refs,
                                GDateTime            **out_summary_timestamp,
                                GCancellable          *cancellable,
                                GError               **error)
{
  g_autoptr(OstreeRepo) repo = NULL;
  g_auto(OstreeCollectionRefv) refs = NULL;
  g_autoptr(GDateTime) summary_timestamp = NULL;

  if (!ostree_sysroot_get_repo (sysroot,
                                &repo,
                                cancellable,
                                error))
    return FALSE;

  if (!get_refs (repo, &refs, cancellable, error))
    return FALSE;

  if (!get_summary_timestamp (repo, &summary_timestamp, cancellable, error))
    return FALSE;

  g_assert (out_refs);
  g_assert (out_summary_timestamp);
  *out_refs = g_steal_pointer (&refs);
  *out_summary_timestamp = g_steal_pointer (&summary_timestamp);
  return TRUE;
}

static gboolean
update_service_file (gboolean       advertise_updates,
                     const gchar   *avahi_service_directory,
                     GCancellable  *cancellable,
                     GError       **error)
{
  g_autoptr(OstreeSysroot) sysroot = NULL;
  g_autofree gchar *commit_checksum = NULL;
  g_autofree gchar *commit_ostree_path = NULL;
  guint64 commit_timestamp;
  g_autoptr(GDateTime) commit_date_time = NULL;
  g_autofree gchar *formatted_commit_date_time = NULL;
  gboolean delete;
  g_autoptr(GError) local_error = NULL;

  g_return_val_if_fail (avahi_service_directory != NULL, FALSE);
  g_return_val_if_fail (cancellable == NULL || G_IS_CANCELLABLE (cancellable),
                        FALSE);
  g_return_val_if_fail (error == NULL || *error == NULL, FALSE);

  /* Work out what commit we would advertise. Errors here are not fatal, as we
   * want to delete the file on failure. */
  sysroot = ostree_sysroot_new_default ();

  if (!ostree_sysroot_load (sysroot, cancellable, &local_error))
    {
      g_warning ("Error loading sysroot: %s", local_error->message);
      g_clear_error (&local_error);
    }
  else if (!eos_updater_sysroot_get_advertisable_commit (sysroot,
                                                         &commit_checksum,
                                                         &commit_ostree_path,
                                                         &commit_timestamp,
                                                         &local_error))
    {
      g_warning ("Error getting advertisable commit: %s", local_error->message);
      g_clear_error (&local_error);
    }

  if (commit_checksum != NULL)
    {
      commit_date_time = g_date_time_new_from_unix_utc ((gint64) commit_timestamp);
      formatted_commit_date_time = g_date_time_format (commit_date_time,
                                                       "%FT%T%:z");
    }

  /* Work out the update policy. */
  if (!advertise_updates && commit_checksum != NULL)
    {
      g_message ("Advertising updates is disabled. Deployed commit ‘%s’ "
                 "(%s, %s) will not be advertised.",
                 commit_checksum, formatted_commit_date_time,
                 commit_ostree_path);
      delete = TRUE;
    }
  else if (advertise_updates && commit_checksum == NULL)
    {
      g_message ("Advertising updates is enabled, but no appropriate deployed "
                 "commits were found. Not advertising updates.");
      delete = TRUE;
    }
  else if (!advertise_updates && commit_checksum == NULL)
    {
      g_message ("Advertising updates is disabled, and no appropriate deployed "
                 "commits were found. Not advertising updates.");
      delete = TRUE;
    }
  else
    {
      g_message ("Advertising updates is enabled, and deployed commit ‘%s’ "
                 "(%s, %s) will be advertised.", commit_checksum,
                 formatted_commit_date_time, commit_ostree_path);
      delete = FALSE;
    }

  /* Delete the old format service file, regardless of the current state. */
    {
      g_autoptr(GFile) old_service_file = NULL;
      g_autofree gchar *old_service_file_path = NULL;

      old_service_file_path = g_build_filename (avahi_service_directory,
                                                "eos-updater.service", NULL);
      old_service_file = g_file_new_for_path (old_service_file_path);

      if (!g_file_delete (old_service_file, cancellable, &local_error) &&
          !g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
        {
          g_debug ("Error deleting old Avahi service file (‘%s’); ignoring: %s",
                   old_service_file_path, local_error->message);
        }

      g_clear_error (&local_error);
    }

  /* Apply the policy. */
  if (!delete)
    {
      g_autoptr(GDateTime) summary_timestamp = NULL;
      g_auto(OstreeCollectionRefv) refs = NULL;

      if (!get_refs_and_summary_timestamp (sysroot,
                                           &refs,
                                           &summary_timestamp,
                                           cancellable,
                                           error))
        return FALSE;

      if (!eos_ostree_avahi_service_file_generate (avahi_service_directory,
                                                   refs,
                                                   summary_timestamp,
                                                   NULL, /* no options, use defaults */
                                                   cancellable,
                                                   error))
        {
          eos_ostree_avahi_service_file_delete (avahi_service_directory,
                                                0,
                                                cancellable,
                                                NULL);
          return FALSE;
        }

      return TRUE;
    }
  else
    {
      return eos_ostree_avahi_service_file_delete (avahi_service_directory,
                                                   0,
                                                   cancellable,
                                                   error);
    }
}

/* main() exit codes. */
enum
{
  EXIT_OK = EXIT_SUCCESS,
  EXIT_FAILED = 1,
  EXIT_INVALID_ARGUMENTS = 2,
  EXIT_BAD_CONFIGURATION = 3,
};

static int
fail (gboolean     quiet,
      int          exit_status,
      const gchar *error_message,
      ...) G_GNUC_PRINTF (3, 4);

static int
fail (gboolean     quiet,
      int          exit_status,
      const gchar *error_message,
      ...)
{
  va_list ap;
  g_autofree gchar *formatted_message = NULL;

  g_return_val_if_fail (exit_status > 0, EXIT_FAILED);

  if (quiet)
    return exit_status;

  /* Format the arguments. */
  va_start (ap, error_message);
  formatted_message = g_strdup_vprintf (error_message, ap);
  va_end (ap);

  /* Include the usage. */
  g_printerr ("%s: %s\n", g_get_prgname (), formatted_message);

  return exit_status;
}

static int
usage (GOptionContext *context,
       gboolean        quiet,
       const gchar    *error_message,
       ...) G_GNUC_PRINTF (3, 4);

static int
usage (GOptionContext *context,
       gboolean        quiet,
       const gchar    *error_message,
       ...)
{
  va_list ap;
  g_autofree gchar *formatted_message = NULL;
  g_autofree gchar *help = NULL;

  if (quiet)
    return EXIT_INVALID_ARGUMENTS;

  /* Format the arguments. */
  va_start (ap, error_message);
  formatted_message = g_strdup_vprintf (error_message, ap);
  va_end (ap);

  /* Include the usage. */
  help = g_option_context_get_help (context, TRUE, NULL);
  g_printerr ("%s: %s\n\n%s\n", g_get_prgname (), formatted_message, help);

  return EXIT_INVALID_ARGUMENTS;
}

int
main (int    argc,
      char **argv)
{
  g_autoptr(GError) error = NULL;
  gboolean advertise_updates = FALSE;
  g_autoptr(GOptionContext) context = NULL;
  g_autofree gchar *avahi_service_directory = NULL;
  g_autofree gchar *config_file = NULL;
  gboolean quiet = FALSE;

  const GOptionEntry entries[] =
    {
      { "service-directory", 'd',
        G_OPTION_FLAG_NONE, G_OPTION_ARG_FILENAME, &avahi_service_directory,
        "Directory containing Avahi .service files "
        "(default: " SYSCONFDIR "/avahi/services" ")",
        "PATH" },
      { "config-file", 'c',
        G_OPTION_FLAG_NONE, G_OPTION_ARG_FILENAME, &config_file,
        "Configuration file to use (default: "
        SYSCONFDIR "/eos-updater/eos-update-server.conf" ")", "PATH" },
      { "quiet", 'q', G_OPTION_FLAG_NONE, G_OPTION_ARG_NONE, &quiet,
        "Do not print anything; check exit status for success", NULL },
      { NULL }
    };

  setlocale (LC_ALL, "");

  /* Command line parsing and defaults. */
  context = g_option_context_new ("— Endless OS Avahi Advertisement Updater");
  g_option_context_add_main_entries (context, entries, NULL);
  g_option_context_set_summary (context,
                                "Update the Avahi DNS-SD service advertisement "
                                "for advertising updates from this machine to "
                                "the local network, enabling or disabling it "
                                "as appropriate to match the current "
                                "configuration and OSTree state.");

  if (!g_option_context_parse (context,
                               &argc,
                               &argv,
                               &error))
    {
      return usage (context, quiet, "Failed to parse options: %s",
                    error->message);
    }

  if (avahi_service_directory == NULL)
    avahi_service_directory = g_strdup (eos_avahi_service_file_get_directory ());

  /* Load our configuration. */
  if (!eus_read_config_file (config_file, &advertise_updates, NULL, &error))
    {
      return fail (quiet, EXIT_BAD_CONFIGURATION,
                   "Failed to load configuration file: %s", error->message);
    }

  /* Update the Avahi configuration file to match. */
  if (!update_service_file (advertise_updates, avahi_service_directory, NULL,
                            &error))
    {
      return fail (quiet, EXIT_FAILED,
                   "Failed to update service file: %s", error->message);
    }

  return EXIT_OK;
}
